### DISCUSSION TITLE
Implementing Redo-Undo Feature in glsp's graph

### DISCUSSION THREAD

**Discussion by Adem-Kri** [eclipse-glsp/glsp#1172] [Q&A] [answered]
Hello,
I'm developing a custom graph editor based on the minimal project (theia-node-json). 
I've added various features to my project, and now I want to implement the **undo-redo** feature. However, I'm not sure how to do this effectively.

I've reviewed previous discussions, including #983, but haven't found solutions that match my project's requirements

Thank you in advance for your assistance!

**Comment by tortmayr** [comment, marked as answer]
Hi @Adem-Kri,
what GLSP version are you using? With 2.0.0 the operation handling has been revamped and undo-redo support is now provided out of the box. OperationHandlers now wrap their execution logic in a `Command` which can then be executed on a command stack that supports undo/redo.

The `node-json-theia` project template has been updated, so maybe you could have a look at e.g. the [`CreateTaskHandler`](https://github.com/eclipse-glsp/glsp-examples/blob/master/project-templates/node-json-theia/tasklist-glsp-server/src/handler/create-task-node-handler.ts).
For json-based source models we provide reusable base implementations (`JsonOperationHandler, `JsonCreateNodeOperationHandler`, `JsonCreateEdgeOperationHandler`). Simply wrap the changes you make to the source model in `this.commandOf` and it will create a command that comprises all changes made to the model.
